File: ts/src/typed.ts
/**
 * Distinct string domains of the public API get distinct branded types.
 * Untyped string literals still convert at call sites via the constructor
 * helpers, but a value of one domain cannot silently cross into another.
 */

import { create } from "@bufbuild/protobuf";
import semver from "semver";
import type { SchemaIdentity } from "./gen/schemapb/schema_pb.js";
import { SchemaIdentitySchema } from "./gen/schemapb/schema_pb.js";

declare const brand: unique symbol;

type Branded<B extends string> = string & { readonly [brand]: B };

export type Namespace = Branded<"Namespace">;
export type SchemaName = Branded<"SchemaName">;
export type FieldName = Branded<"FieldName">;
export type DefName = Branded<"DefName">;
export type TemplateName = Branded<"TemplateName">;
export type RuleId = Branded<"RuleId">;
export type GroupName = Branded<"GroupName">;
export type VariantKey = Branded<"VariantKey">;
export type Format = Branded<"Format">;

/** Core string formats of the spec's format registry. */
export const FORMATS = {
  email: "email" as Format,
  url: "url" as Format,
  uuid: "uuid" as Format,
  ipv4: "ipv4" as Format,
  ipv6: "ipv6" as Format,
  ip: "ip" as Format,
  hostname: "hostname" as Format,
  date: "date" as Format,
  time: "time" as Format,
  datetime: "datetime" as Format,
} as const;

/**
 * Version is an opaque, always-valid semver value: either zero
 * ("unversioned", serializes to "") or a canonical semver produced by
 * parseVersion / mustVersion / ver. Construction outside this module is
 * impossible — the class constructor is private.
 */
export class Version {
  static readonly zero = new Version("");

  readonly #s: string;

  private constructor(s: string) {
    this.#s = s;
  }

  /** Builds a release version ("v1.2.3"). */
  static of(major: number, minor: number, patch: number): Version {
    return new Version(`v${major}.${minor}.${patch}`);
  }

  /**
   * Parses a semver string ("v1.2.3", "1.2.3-rc.1"; shorthands
   * canonicalize). An empty string is the zero Version.
   */
  static parse(s: string): Version {
    if (s === "") {
      return Version.zero;
    }
    const normalized = s.startsWith("v") ? s.slice(1) : s;
    const parsed = semver.parse(
      semver.coerce(normalized, { includePrerelease: true }) ?? normalized,
    );
    if (parsed === null) {
      throw new Error(`schemapb: invalid version ${JSON.stringify(s)}`);
    }
    return new Version(`v${parsed.version}`);
  }

  get isZero(): boolean {
    return this.#s === "";
  }

  /** Canonical wire form ("" when unversioned). */
  toString(): string {
    return this.#s;
  }

  /** Semver precedence (-1, 0, 1); the zero Version sorts first. */
  compare(o: Version): number {
    if (this.#s === o.#s) {
      return 0;
    }
    if (this.#s === "") {
      return -1;
    }
    if (o.#s === "") {
      return 1;
    }
    return semver.compare(this.#s.slice(1), o.#s.slice(1));
  }
}

/**
 * Builds a schema identity. Declare it ONCE next to the schema and reuse the
 * same value everywhere the identity is needed (newSchema, refId,
 * registry.get) — a typo then cannot happen twice.
 */
export function id(ns: string, name: string, ver: Version): SchemaIdentity {
  return create(SchemaIdentitySchema, {
    namespace: ns,
    name,
    version: ver.toString(),
  });
}

File: rust/src/typed.rs
//! Typed identifier domains (newtypes) and the opaque semver [`Version`].
//! An invalid version is unrepresentable (principle 2); validation and
//! ordering delegate to the `semver` crate.

use crate::gen::schemapb::SchemaIdentity;

macro_rules! string_domain {
    ($(#[$doc:meta] $name:ident),+ $(,)?) => {
        $(
            #[$doc]
            #[derive(Debug, Clone, PartialEq, Eq, Hash)]
            pub struct $name(pub String);

            impl From<&str> for $name {
                fn from(s: &str) -> Self {
                    Self(s.to_owned())
                }
            }

            impl std::fmt::Display for $name {
                fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
                    f.write_str(&self.0)
                }
            }
        )+
    };
}

string_domain!(
    /// A schema identity's grouping namespace.
    Namespace,
    /// A schema identity's name, unique within its namespace.
    SchemaName,
    /// One field's name inside a schema.
    FieldName,
    /// A reusable sub-schema name in the root defs map.
    DefName,
    /// A render template name carried by a schema.
    TemplateName,
    /// The stable id of an authored validation rule.
    RuleId,
    /// A field's informative section label.
    GroupName,
    /// A `OneOf` variant key (the discriminator value).
    VariantKey,
    /// A string-format registry identifier ("email", "k8s.quantity").
    Format,
);

/// An always-valid semver value; the zero version means "unversioned" and
/// serializes to the empty string. Constructible only through
/// [`Version::of`] / [`Version::parse`].
#[derive(Debug, Clone, PartialEq, Eq, Default)]
pub struct Version(Option<semver::Version>);

impl Version {
    #[must_use]
    pub const fn zero() -> Self {
        Self(None)
    }

    #[must_use]
    pub const fn of(major: u64, minor: u64, patch: u64) -> Self {
        Self(Some(semver::Version::new(major, minor, patch)))
    }

    /// Parses "v1.2.3" / "1.2.3-rc.1"; the empty string is the zero version.
    pub fn parse(s: &str) -> Result<Self, semver::Error> {
        if s.is_empty() {
            return Ok(Self(None));
        }
        let trimmed = s.strip_prefix('v').unwrap_or(s);
        semver::Version::parse(trimmed).map(|v| Self(Some(v)))
    }

    #[must_use]
    pub const fn is_zero(&self) -> bool {
        self.0.is_none()
    }

    /// Semver precedence; the zero version sorts before any real version.
    #[must_use]
    pub fn compare(&self, other: &Self) -> std::cmp::Ordering {
        match (&self.0, &other.0) {
            (None, None) => std::cmp::Ordering::Equal,
            (None, Some(_)) => std::cmp::Ordering::Less,
            (Some(_), None) => std::cmp::Ordering::Greater,
            (Some(a), Some(b)) => a.cmp(b),
        }
    }
}

impl std::fmt::Display for Version {
    /// Canonical wire form ("v1.2.3", "" when unversioned).
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        self.0.as_ref().map_or(Ok(()), |v| write!(f, "v{v}"))
    }
}

/// Builds a schema identity handle: declare it once next to the schema and
/// reuse the same value everywhere the identity is needed.
#[must_use]
pub fn id(ns: &str, name: &str, ver: &Version) -> SchemaIdentity {
    SchemaIdentity {
        namespace: ns.to_owned(),
        name: name.to_owned(),
        version: ver.to_string(),
    }
}
